custom/icds_reports/sqldata/exports/dashboard_usage.py

```diff
@@ -0,0 +1,209 @@
+        excel_rows = []
+        filters = [['Generated at', india_now()]]
+        headers = ['Sr.No', 'State/UT Name', 'District Name', 'Block Name', 'Username', 'Level', 'Role',
+                   'Launched?', 'Last Login', 'Logged in the last week?', 'Total', 'Child', 'Pregnant Women',
+                   'Demographics', 'System Usage', 'AWC infrastructure', 'Child Growth Monitoring List',
+                   'ICDS - CAS Monthly Register', 'AWW Performance Report', 'LS Performance Report',
+                   'Take Home Ration']
+        excel_rows.append(headers)
+        serial_count = 0
+        for user_location in self.user.get_sql_locations(self.domain):
+            # getting the location types to retrieve for this user location
+            location_type_filter = {
+                self.get_location_id_string_from_location_type(user_location.location_type_name):
+                    user_location.get_id}
+            all_awc_locations = AwcLocation.objects.filter(**location_type_filter).values(*self.required_fields)
+            # converting the result set to matrix to fetch ancestors for a given location
+            location_matrix, location_ids =\
+                self.convert_rs_to_matrix(all_awc_locations, self.get_location_id_string_from_location_type(
+                    user_location.location_type_name))
+            users = self.get_users_by_location(location_ids)
+            records = ICDSAuditEntryRecord.objects.filter(url='/a/icds-dashboard-qa/cas_export')\
+                .annotate(indicator=KeyTextTransform('indicator', 'post_data')).values('indicator',
+                                                                                       'username')\
+                .annotate(count=Count('indicator')).order_by('username', 'indicator')
+            self.prepare_is_launched_agg_list(user_location.location_type_name, user_location.get_id)
+            user_counts = defaultdict(int)
+            user_indicators = defaultdict(list)
+            for record in records:
+                user_counts[record['username']] += record['count']
+                user_indicators[record['username']].append(record['count'])
+            # accumulating the indicator counts
+            for user in users:
+                indicator_count = user_indicators[user['username']]
+                user_sql_location_ids = user['assigned_location_ids']
+                if isinstance(user_sql_location_ids, str):
+                    user_sql_location_ids = [user_sql_location_ids]
+
+                for user_sql_location in user_sql_location_ids:
+                    # getting the location type to look up in matrix
+                    location_type_id = self.sql_locations[user_sql_location]
+                    column_index = self.location_types.index(location_type_id)
+                    user_location_row = None
+                    # iterating and getting the db row from matrix
+                    for row in location_matrix:
```

Comment:
Row on this matrix will contain entire hierarchy right from state to awc?
then its not required. Because there is dashboard user which is at awc level.